daml-lf/spec/daml-lf-1.rst

```diff
@@ -2569,6 +2575,150 @@ Map functions
+
+  This raises an error if the key is not a valid map key. Keys are
+  compared according to the rules listed below.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_LOOKUP : ∀ α. ∀ β.  α → 'GenMap' α β → 'Optional' α``
+
+  Lookups the value at a key in the map.
+
+  This raises an error if the key is not a valid map key. Keys are
+  compared according to the rules listed below.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_DELETE : ∀ α. ∀ β.  α → 'GenMap' α β → 'GenMap' α β``
+
+  Deletes a key and its value from the map. When the key is not a
+  member of the map, the original map is returned.
+
+  This raises an error if the key is not a valid map key. Keys are
+  compared according to the rules listed below.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_LIST : ∀ α. ∀ β.  'GenMap' α β → 'List' ⟨ key: α, value: β ⟩``
+
+  Converts to a list of key/value pairs. Unlike ``'MAP_LIST'``, the output
+  list is guaranteed to be in the order that the keys were first inserted.
+
+  [*Available in versions >= 1.dev*]
+
+* ``GENMAP_SIZE : ∀ α. ∀ β.  'GenMap' α β → 'Int64'``
+
+  Return the number of elements in the map.
+
+  [*Available in versions >= 1.dev*]
+
+
+**Validity of Keys:** A key is valid if and only if it is equivalent
+to itself according to the relation ``~ᵥ`` defined below. Attempts to
+use an invalid key in the operations above always result in a runtime
+error.
+
+Of particular note, the following values are never valid keys:
+
+* Lambda expressions ``λ x : τ . e``
+* Type abstractions ``Λ α : k . e``
+* (Partially applied) built-in functions
+* (Un-evaluated) Update/Scenario values
+* Any value containing an invalid key
+
+**Comparison of Keys:** We define here the relation ``~ᵥ`` on value
+expressions that is used for key comparison. This is a partial
+equivalence relation over all values, but a (total) equivalence
+relation over valid keys.
+
+This relation is not exposed as a builtin function, but it coincides
+with the built-in equality for any given type, if that built-in isdefined. ::
+
+                                  ┌──────────┐
+  Generic Equivalence Relation    │ e₁ ~ᵥ e₂ │
+                                  └──────────┘
+
+  LitInt64₁ and LitInt64₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitInt64
+  LitInt64₁ ~ᵥ LitInt64₂
+
+  LitNumeric₁ and LitNumeric₂ have the same scale and value
+  ——————————————————————————————————————————————————— GenEqLitNumeric
+  LitNumeric₁ ~ᵥ LitNumeric₂
+
+  LitText₁ and LitText₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitText
+  LitText₁ ~ᵥ LitText₂
+
+  LitDate₁ and LitDate₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitDate
+  LitDate₁ ~ᵥ LitDate₂
+
+  LitTimestamp₁ and LitTimestamp₂ have the same value
+  ——————————————————————————————————————————————————— GenEqLitTimestamp
+  LitTimestamp₁ ~ᵥ LitTimestamp₂
+
+  cid₁ and cid₂ are the same
```

Comment:
Do we want to allow contract ids? The absolute vs relative thing makes that a bit dangerous so at least on the DAML level I probably wouldn’t expose it.

Reply:
Contract IDs are fine here as long as we stick to the following two assumptions:
(1) the rewriting from relative to absolute IDs is injective and no relative ID is ever rewritten to an absolute ID that is in use already,
(2) the rewriting of all contract IDs in a transaction happens in one go without any map operations performed in the middle.
I wonder if we should document these assumptions (in better words than mine)?

Reply:
Oh I see, I think I didn’t realize how the rewriting works until now. Makes sense now but yeah we should definitely document this imho.